consensus/parlia/snapshot.go

```diff
@@ -267,8 +268,19 @@ func (s *Snapshot) apply(headers []*types.Header, chain consensus.ChainHeaderRea
 			}
 		}
 		snap.Recents[number] = validator
+		snap.RecentForkHashes[number] = hex.EncodeToString(header.Extra[extraVanity-nextForkHashSize : extraVanity])
+		snap.updateAttestation(header, chainConfig, s.config)
 		// change validator set
 		if number > 0 && number%s.config.Epoch == uint64(len(snap.Validators)/2) {
+			epochKey := math.MaxUint64 - header.Number.Uint64()/s.config.Epoch // impossible used as a block number
```

Comment:
"math.MaxUint64" is to avoid conflict with "block number", right?
why not just create another element in Snapshot structure, like "RecentsEpochs", similar to "RecentForkHashes"?

Reply:
another corner case: suppose validator size grow from 7 -> 11?
1.a node passed block height: 203, "snap.Recents[epochKey]" was set
2.it passed 204 and got restarted
3.after restart, it start to sync from 205, but `snap.Recents[epochKey]` was cleared. when it hit block height: 205, it will set "snap.Recents[epochKey]" one more time.

is any potential risk that "recents" got cleared twice in this corner case?

Reply:
> another corner case: suppose validator size grow from 7 -> 11? 1.a node passed block height: 203, "snap.Recents[epochKey]" was set 2.it passed 204 and got restarted 3.after restart, it start to sync from 205, but `snap.Recents[epochKey]` was cleared. when it hit block height: 205, it will set "snap.Recents[epochKey]" one more time.
> 
> is any potential risk that "recents" got cleared twice in this corner case?

when restart
will get a snapshot from the disk firstly, then apply headers

Reply:
> "math.MaxUint64" is to avoid conflict with "block number", right? why not just create another element in Snapshot structure, like "RecentsEpochs", similar to "RecentForkHashes"?

fixed

Reply:
> "math.MaxUint64" is to avoid conflict with "block number", right? why not just create another element in Snapshot structure, like "RecentsEpochs", similar to "RecentForkHashes"?

changed